### PR TITLE
Multicall support for the public resolver

### PR DESCRIPTION
**Pull request by Arachnid**

### CODE REVIEW

File: contracts/PublicResolver.sol [hunk 1]
```diff
@@ -1,4 +1,5 @@
 pragma solidity ^0.5.0;
+pragma experimental ABIEncoderV2;
```
Comment (decanus) [inline contracts/PublicResolver.sol]: I don't like using experimental, I really don't.

File: contracts/PublicResolver.sol [hunk 2]
```diff
@@ -52,4 +53,14 @@ contract PublicResolver is ABIResolver, AddrResolver, ContentHashResolver, DNSRe
         address owner = ens.owner(node);
         return owner == msg.sender || authorisations[node][owner][msg.sender];
     }
+
+    function multicall(bytes[] calldata datas) external returns(bytes[] memory rets) {
```
Comment (decanus) [inline contracts/PublicResolver.sol]: the plural of data is data.

File: contracts/PublicResolver.sol (hunk 1 above)
Comment (Arachnid) [reply contracts/PublicResolver.sol]: Nor do I, but it's the only way to implement this. It's a fairly basic use of it, too, so I think it's safe.

File: contracts/PublicResolver.sol (hunk 2 above)
Comment (Arachnid) [reply contracts/PublicResolver.sol]: Yeah, but then we have no variable name to use for the singular.

Comment (decanus) [reply contracts/PublicResolver.sol]: I don't see dada used anywehre?

Comment (Arachnid) [reply contracts/PublicResolver.sol]: But if we did I'd have to rename this one.

Is this important?

Comment (makoto) [reply contracts/PublicResolver.sol]: Singular of `data` is `datum`

Comment (makoto) [reply contracts/PublicResolver.sol]: or you can call `arrayofdata` a bit long though

File: contracts/PublicResolver.sol [hunk 3]
```diff
@@ -52,4 +53,14 @@ contract PublicResolver is ABIResolver, AddrResolver, ContentHashResolver, DNSRe
...
+        rets = new bytes[](datas.length);
+        for(uint i = 0; i < datas.length; i++) {
+            (bool success, bytes memory ret) = address(this).delegatecall(datas[i]);
+            require(success);
```
Comment (makoto) [inline contracts/PublicResolver.sol]: What is our policy of adding error message as a second parameter? It's inconsistent so far.

File: contracts/PublicResolver.sol [hunk 4]
```diff
@@ -52,4 +53,14 @@ contract PublicResolver is ABIResolver, AddrResolver, ContentHashResolver, DNSRe
         address owner = ens.owner(node);
         return owner == msg.sender || authorisations[node][owner][msg.sender];
     }
+
+    function multicall(bytes[] calldata datas) external returns(bytes[] memory rets) {
+        rets = new bytes[](datas.length);
```
Comment (makoto) [inline contracts/PublicResolver.sol]: Please use more descriptive variable names. I don't know what `rets` stands for. Did you get inspiration from MakerDAO?

File: test/TestPublicResolver.js [hunk 5]
```diff
@@ -659,6 +659,40 @@ contract('PublicResolver', function (accounts) {
...
+          assert.equal(tx.logs[1].args.a, accounts[1]);
+          assert.equal(tx.logs[2].event, "TextChanged");
+          assert.equal(tx.logs[2].args.node, node);
+          assert.equal(tx.logs[2].args.key, "url");
+
+          assert.equal(await resolver.methods['addr(bytes32)'](node), accounts[1]);
+          assert.equal(await resolver.text(node, "url"), "https://ethereum.org/");
+        });
+
+        // Disabled because web3 doesn't have a function for decoding the return ABI
```
Comment (makoto) [inline test/TestPublicResolver.js]: You could try decoding with ethersjs which I think it does support.

```Meta-classes create JavaScript objects from any contract ABI, including ABIv2 and Human-Readable ABI```

https://docs.ethers.io/ethers.js/html/index.html?highlight=abiv2
https://blog.ricmoo.com/solidity-abiv2-a-foray-into-the-experimental-a6afd3d47185

Comment (Arachnid) [reply test/TestPublicResolver.js]: I don't really want to include a new library just for this one test. I'll work around it.

File: contracts/PublicResolver.sol (hunk 3 above)
Comment (Arachnid) [reply contracts/PublicResolver.sol]: I've been omitting it, because error messages cause significant bloat.

**Comment by Arachnid** [comment]
@decanus @makoto PTAL.

**Comment by makoto** [comment]
For some reason when I run `yarn prepublishOnly`, the artifacts only contain abi of `setName`.

```
~/.../migration/resolvers (multicall)$!more
more build/contracts/Resolver.json
{
  "contractName": "Resolver",
  "abi": [
    {
      "constant": false,
      "inputs": [
        {
          "name": "node",
          "type": "bytes32"
        },
        {
          "name": "name",
          "type": "string"
        }
      ],
      "name": "setName",
      "outputs": [],
      "payable": false,
      "stateMutability": "nonpayable",
      "type": "function"
    }
  ],
```

However when I add some random functions (eg: `function foo() external`), then it generates full abi. Can you check if it happens to you as well? @Arachnid

**Pull request merged**